### PR TITLE
feat: pagination to typescript

### PR DESCRIPTION
**Pull request by DorianMaliszewski**
## Type

- Refactor

#### What is expected?

Migrate pagination to typescript

### CODE REVIEW

File: src/components/Pagination/__tests__/__snapshots__/index.test.tsx.snap [hunk 1]
```diff
@@ -1,24 +1,7 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`Pagination should render correctly component 1`] = `
+exports[`Pagination should render correctly PaginationContainer 1`] = `
```
Comment (chambo-e) [inline src/components/Pagination/__tests__/__snapshots__/index.test.tsx.snap:3]: All these changes kinda scares me 😄 

To be able to better see that nothing changed, could you split your PR in two commits, one with the conversion to ts and one with the new tests ?

File: src/components/Pagination/index.tsx [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+  canLoadMore: boolean
+  data?: unknown[] | null
+} & UsePaginationReturn<unknown>
+
+export type PaginationState<T = unknown> = {
+  canLoadMore: boolean
+  data?: T[] | null
+} & UsePaginationReturn<T>
+
+const PaginationContext = createContext({} as PaginationContextValue)
```
Comment (chambo-e) [inline src/components/Pagination/index.tsx:34]:
```suggestion
const PaginationContext = createContext<PaginationContextValue>({})
```

Maybe ?

File: src/components/Pagination/index.tsx [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+        disabled={page === maxPage || isLoadingPage}
+        onClick={goToLastPage}
+      >
+        Last
+      </Button>
+    </div>
+  )
+}
+
+const DefaultRightComponent: FunctionComponent<PaginationComponentProps> = ({
```
Comment (chambo-e) [inline src/components/Pagination/index.tsx:139]:
```suggestion
const DefaultRightComponent: VoidFunctionComponent<PaginationComponentProps> = ({
```

File: src/components/Pagination/index.tsx [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+  paginationState: { page },
+}) => <div>Current : {page}</div>
+
+const DefaultLoaderComponent: FunctionComponent<PaginationComponentProps> =
```
Comment (chambo-e) [inline src/components/Pagination/index.tsx:143]:
```suggestion
const DefaultLoaderComponent: VoidFunctionComponent <PaginationComponentProps> =
```

File: src/components/Pagination/index.tsx [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+  RightComponent: FunctionComponent<PaginationComponentProps<T>>
+  LeftComponent: FunctionComponent<PaginationComponentProps<T>>
+  PaginationContainer: FunctionComponent<PaginationComponentProps<T>>
+  MiddleComponent: FunctionComponent<PaginationComponentProps<T>>
+  LoaderComponent: FunctionComponent<PaginationComponentProps<T>>
+}
+
+const Pagination = forwardRef(PaginationForwardFn) as PaginationType
+
+Pagination.propTypes = {
```
Comment (chambo-e) [inline src/components/Pagination/index.tsx:420]: Put the propTypes on the component, not the forwardRef function :)

File: src/components/Pagination/index.tsx [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
+  // eslint-disable-next-line react/forbid-prop-types
+  data: PropTypes.arrayOf(() => null),
```
Comment (chambo-e) [inline src/components/Pagination/index.tsx:423]: PropTypes.any does not work anymore ?

**Review by matthprost — CHANGES_REQUESTED** [review]
Screenshot 2021-09-06 at 11 09 50

No more prop definition in argsTable 😢

File: src/components/Pagination/index.tsx (hunk 6 above)
Comment (DorianMaliszewski) [reply src/components/Pagination/index.tsx:423]: Not for unknown 😢

File: src/components/Pagination/index.tsx (hunk 5 above)
Comment (DorianMaliszewski) [reply src/components/Pagination/index.tsx:420]: According to https://reactjs.org/docs/typechecking-with-proptypes.html : proptypes must be defined on the forwardRef return (ForwardRefExoticComponent interface include propTypes and defaultProps). To make sure it was right, i tried to put propTypes on PaginationForwardFn and the ts compiler throw me an error. I think i will keep propTypes on forwardRef return. 😅

File: src/components/Pagination/__tests__/__snapshots__/index.test.tsx.snap (hunk 1 above)
Comment (DorianMaliszewski) [reply src/components/Pagination/__tests__/__snapshots__/index.test.tsx.snap:3]: I put it at the end of the tests, like this the changes is more comprehensive.

File: src/components/Pagination/index.tsx (hunk 2 above)
Comment (DorianMaliszewski) [reply src/components/Pagination/index.tsx]: Doesn't work need to cast the object to the desired type :/

File: src/components/Pagination/index.tsx [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+  RightComponent?: FunctionComponent<PaginationComponentProps<T>>
+  LeftComponent?: FunctionComponent<PaginationComponentProps<T>>
+  MiddleComponent?: FunctionComponent<PaginationComponentProps<T>>
+}
+
+const PaginationContainer: VoidFunctionComponent<PaginationContainerProps> = ({
+  LeftComponent = DefaultLeftComponent,
+  MiddleComponent = DefaultMiddleComponent,
+  RightComponent = DefaultRightComponent,
+  paginationState,
```
Comment (vincentaudebert) [inline src/components/Pagination/index.tsx:174]: Is it not missing in the props?

**Review by matthprost — CHANGES_REQUESTED** [review]

File: src/components/Pagination/__tests__/index.test.tsx [hunk 8]
```diff
@@ -19,7 +20,13 @@ const ExampleChildren = () => {
   )
 }
 
-const loadMore = async ({ page, perPage }) => {
+const loadMore = async ({
+  page,
+  perPage = 0,
+}: {
+  page: number
+  perPage?: number
```
[the remark this answers is not in the document]
Comment (DorianMaliszewski) [reply src/components/Pagination/__tests__/index.test.tsx:28]: By default we only want one page ?

**Review by matthprost — APPROVED** [review]

File: src/components/List/__tests__/__snapshots__/index.js.snap [hunk 9]
```diff
@@ -26636,11 +26635,6 @@ exports[`List should render correctly with pagination and page loading 1`] = `
   vertical-align: middle;
 }
 
-.cache-1okuwm0.cache-1okuwm0 {
```
Comment (chambo-e) [inline src/components/List/__tests__/__snapshots__/index.js.snap:26639]: Intended ?

File: src/components/Pagination/getPageNumbers.ts [hunk 10]
```diff
@@ -1,4 +1,15 @@
-export default function getPageNumbers(currentPage, pageCount, range = 5) {
+/**
+ * Regturn a list of page numbers around the currentPage
```
Comment (chambo-e) [inline src/components/Pagination/getPageNumbers.ts:2]:
```suggestion
 * Return a list of page numbers around the currentPage
```

File: src/components/List/__tests__/__snapshots__/index.js.snap (hunk 9 above)
Comment (DorianMaliszewski) [reply src/components/List/__tests__/__snapshots__/index.js.snap:26639]: Yes it was a misconception

**Review by chambo-e — APPROVED** [review]

**Pull request merged**